wagtail/core/query.py

```diff
@@ -237,6 +272,71 @@ def not_public(self):
         """
         return self.exclude(self.public_q())
 
+    def _get_closest_view_restriction(self, page):
+        """
+        Return the PageViewRestriction most applicable to the provided ``page`` instance.
+
+        View restrictions apply to the selected page AND their descendants, so for each
+        page, it's the restriction applied to the 'closest' ancestor that determines
+        what restrictions apply to it.
+        """
+        for vr in self.all_view_restrictions:
+            if page.id == vr.page.id or page.is_descendant_of(vr.page):
+                return vr
+
+    def for_user(self, user=None, include_password_protected=False):
+        """
+        Return a queryset of public and private pages that should be
+        accessible to the provided ``user``, according to their group
+        memberships and current page view restriction settings.
+
+        It is impossible to know whether the user knows the password
+        to access a password-protected page/section, so those pages are
+        excluded by default. However, you can use the
+        ``include_password_protected`` keyword argument to override this
+        behaviour and include them if needed.
+        """
+        from wagtail.core.models import PageViewRestriction
+
+        private_page_pks = []
+
+        # assume user is not authenticated
+        user_group_pks = ()
+        non_restrictive_types = [PageViewRestriction.NONE]
+        if include_password_protected:
+            non_restrictive_types.append(PageViewRestriction.PASSWORD)
+
+        # widen criteria for authenticated users
+        if user is not None and user.is_authenticated:
+            user_group_pks = user.groups.all().values_list("pk", flat=True)
+            non_restrictive_types.append(PageViewRestriction.LOGIN)
+
+        # NOTE: we only need the id plus values used by Page.is_descendant_of()
+        for page in self.not_public().only("pk", "path", "depth").distinct().iterator():
+
+            # get the most relevant restriction for this page
+            closest_restriction = self._get_closest_view_restriction(page)
```

Comment:
`.not_public()` also performs a query to get all view restrictions so it might be good to make it only do this once

Reply:
Ah yes. As you can probably tell from `get_view_restrictions()`, I was trying to make it so that existing users of `public()` and `not_public()` weren't penalised by the extra query to prefetch the groups. That's probably a bit over-cautious though, thinking about - and actually means two queries being made here. I feel the 'right' solution is to let `public()` and `not_public()` users take the hit, and keep this simple.